es6-shim.js

```diff
@@ -1241,29 +1241,29 @@
   var MathShims = {
     acosh: function acosh(value) {
       var x = Number(value);
-      if (Number.isNaN(x) || value < 1) { return NaN; }
-      if (x === 1) { return 0; }
-      if (x === Infinity) { return x; }
-      return Math.log(x / Math.E + Math.sqrt(x + 1) * Math.sqrt(x - 1) / Math.E) + 1;
+      if (x * x + 1 === x * x) {
+        return Math.log(x) + Math.log(2);
+      }
+      return Math.log1p(x - 1 + Math.sqrt(x - 1) * Math.sqrt(x + 1));
     },
 
     asinh: function asinh(value) {
       var x = Number(value);
-      if (x === 0 || !globalIsFinite(x)) {
+      if (x === 0) {
         return x;
       }
-      return x < 0 ? -Math.asinh(-x) : Math.log(x + Math.sqrt(x * x + 1));
+      if (x < 0) {
+        return -Math.asinh(-x);
+      }
+      if (x * x + 1 === x * x) {
+        return Math.log(x) + Math.log(2);
+      }
+      return Math.log1p(x + Math.expm1(Math.log1p(x * x) / 2));
```

Comment:
I was just about to post an issue recommending the use of a range-reduction method similar to that used for `Math.acosh` in the current shim:

```javascript
if (x < 0) { return -Math.asinh(-x); }
var sqx = x * x;
if (sqx <= inverseEpsilon - 1) { return Math.log(x + Math.sqrt(sqx + 1)); }
if (sqx <= Number.MAX_VALUE - 1) {
  var sqrtEpsilon = Math.sqrt(Number.EPSILON);
  return Math.log(x * sqrtEpsilon + Math.sqrt(square(x * sqrtEpsilon) + Number.EPSILON)) - Math.log(sqrtEpsilon);
}
var sqrtMax = Math.sqrt(Number.MAX_VALUE);
return Math.log(x / sqrtMax + Math.sqrt(square(x / sqrtMax) + 1 / Number.MAX_VALUE)) + Math.log(Number.MAX_VALUE) / 2;
```

However, this solution looks less hackish (also, I'm not sure whether the function call to `square()` would be worth it if the argument contains an operator, compared to running the inner operations twice).